### PR TITLE
Implement Cuprum Backend Dispatcher With Env Support

### PR DESCRIPTION
**Pull request by leynos**
## Summary
Implements an internal Cuprum backend dispatcher in cuprum/_backend.py that selects between the Rust-backed and pure-Python stream backends at runtime based on the CUPRUM_STREAM_BACKEND environment variable and the availability of the Rust extension. The resolution result is cached for the process lifetime to avoid repeated checks.

Key features:
- Environment-driven selection with three options: auto, rust, python
- Rust availability check cached for performance
- Forced rust mode raises ImportError when the extension is unavailable
- auto mode prefers Rust when available, otherwise falls back to Python
- All logic encapsulated in a new internal module cuprum/_backend.py (no public API changes)

### Core functionality
- New module cuprum/_backend.py containing:
  - StreamBackend (StrEnum) with AUTO, RUST, PYTHON
  - _read_backend_env() -> StreamBackend: reads CUPRUM_STREAM_BACKEND, validates value, defaults to AUTO
  - _check_rust_available() -> bool: cached availability probe for the Rust extension
  - get_stream_backend() -> StreamBackend: main dispatcher implementing precedence rules
  - __all__ = ["StreamBackend", "get_stream_backend"]
- No changes to existing public APIs or public modules; this is an internal dispatcher as planned.

### Tests
- Unit tests: cuprum/unittests/test_backend.py
  - test_stream_backend_enum_values, test_stream_backend_is_str_enum
  - test_auto_returns_rust_when_available, test_auto_returns_python_when_unavailable
  - test_forced_rust_returns_rust_when_available, test_forced_rust_raises_when_unavailable
  - test_forced_python_returns_python, test_invalid_env_var_raises_value_error
  - test_env_var_case_insensitive, test_availability_is_cached, test_cache_clear_allows_recheck
- Behavioural tests scaffolding:
  - tests/features/backend_dispatcher.feature (BDD scenarios for auto, forced rust, forced python)
  - tests/behaviour/test_backend_dispatcher_behaviour.py implementing the steps
- Test helper exposure: cuprum/_testing.py re-exports _check_rust_available for test control

### Documentation
- docs/users-guide.md: note that the dispatcher result is cached for the process lifetime
- docs/roadmap.md: mark 4.2.4 as completed
- docs/execplans/4-2-4-backend-dispatcher.md: new ExecPlan detailing the design, decisions, and validation plan

### Tests and validation
- Autouse fixture clears the rust availability cache between tests to prevent cross-test pollution
- Coverage includes enum behavior, environment-var parsing (case-insensitive, whitespace handling), caching semantics, and ImportError behavior when rust is unavailable
- Behavioural tests mirror unit tests for end-to-end validation
- All changes adhere to project linting/formatting guidelines (no public API changes)

## Rationale
- Keeps the selection logic isolated from the runtime pipelines, allowing future integration without touching core execution code
- Caches the Rust availability probe to improve performance in hot paths
- Provides clear error when a rust-only configuration is requested but the extension is missing
- Maintains backward compatibility: default behavior auto-resolves without requiring changes to callers

## Usage
- Default (no env var): get_stream_backend() resolves to RUST if the extension is available, otherwise PYTHON
- Force Rust: CUPRUM_STREAM_BACKEND=rust will return RUST when available or raise ImportError if not
- Force Python: CUPRUM_STREAM_BACKEND=python will always return PYTHON regardless of availability
- Case-insensitive values and surrounding whitespace are tolerated for env var

Example:
- from cuprum._backend import get_stream_backend, StreamBackend
- backend = get_stream_backend()  # may be StreamBackend.RUST or StreamBackend.PYTHON
- if backend is StreamBackend.RUST: …

## Files touched
- cuprum/_backend.py (new)
- cuprum/unittests/test_backend.py (new)
- tests/features/backend_dispatcher.feature (new)
- tests/behaviour/test_backend_dispatcher_behaviour.py (new)
- cuprum/_testing.py (updated to re-export _check_rust_available)
- docs/users-guide.md (update)
- docs/roadmap.md (update)
- docs/execplans/4-2-4-backend-dispatcher.md (new)

## Acceptance criteria
- [x] Unit tests cover all dispatching scenarios and caching behavior
- [x] Behavioural tests cover end-to-end dispatcher flow
- [x] Documentation updated to reflect caching and usage
- [x] Public API remains unchanged; internal dispatcher does not affect external consumers
- [x] Code formatting and linting conform to project standards

📎 **Task**: https://www.devboxer.com/task/36b1de88-f381-4372-9e1e-cd4bcf6ab28f

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- The `_VALID_VALUES` constant in `cuprum/_backend.py` is currently unused and can be removed to avoid dead code.
- Both the unit tests and behavioural tests define an autouse fixture to clear `_check_rust_available`; consider centralizing this in a shared test helper or `conftest.py` to avoid duplication.

## Individual Comments

### Comment 1
<location> `cuprum/_backend.py:62-66` </location>
<code_context>
+    ValueError
+        If the environment variable contains an unrecognised value.
+    """
+    raw = os.environ.get(_ENV_VAR, "").strip().lower()
+    if not raw:
+        return StreamBackend.AUTO
+    try:
+        return StreamBackend(raw)
+    except ValueError:
+        valid = ", ".join(sorted(v.value for v in StreamBackend))
</code_context>

<issue_to_address>
**issue (bug_risk):** Environment parsing with `StrEnum` + `enum.auto()` means `rust`/`python`/`auto` values will currently be rejected.

Because `StreamBackend` is a `StrEnum` using `enum.auto()`, each member’s value is its *upper‑case* name (`"AUTO"`, `"RUST"`, `"PYTHON"`). Here `raw` is lower‑cased, so `StreamBackend(raw)` will always raise `ValueError` for valid inputs like `"rust"`, and the error message will advertise upper‑case tokens instead, breaking the documented `CUPRUM_STREAM_BACKEND` contract.

To fix this, either:
- map the lower‑case env value via `StreamBackend[raw.upper()]`, or
- stop lowercasing and accept the upper‑case names directly, or
- give the enum explicit lower‑case values (`AUTO = "auto"`, etc.) while keeping the lower‑casing.

Any of these will make env handling accept `rust`/`python`/`auto` as intended and prevent spurious `ValueError`s.
</issue_to_address>

### Comment 2
<location> `cuprum/_backend.py:45` </location>
<code_context>
+    PYTHON = enum.auto()
+
+
+_VALID_VALUES = frozenset(StreamBackend)
+
+
</code_context>

<issue_to_address>
**suggestion:** _VALID_VALUES is currently unused and can either be removed or wired into validation.

If you decide to keep it, consider using `_VALID_VALUES` inside `_read_backend_env` for validation or error messages so its purpose is explicit and the allowed values stay centralised.
</issue_to_address>

### Comment 3
<location> `cuprum/unittests/test_backend.py:113-127` </location>
<code_context>
+        get_stream_backend()
+
+
+@pytest.mark.parametrize(
+    ("env_value", "expected"),
+    [
+        ("AUTO", StreamBackend.PYTHON),
+        ("Rust", StreamBackend.RUST),
+        ("PYTHON", StreamBackend.PYTHON),
+        ("  auto  ", StreamBackend.PYTHON),
+    ],
+    ids=["upper-auto", "mixed-rust", "upper-python", "whitespace-auto"],
+)
+def test_env_var_case_insensitive(
+    monkeypatch: pytest.MonkeyPatch,
+    env_value: str,
+    expected: StreamBackend,
+) -> None:
+    """Env var values are case-insensitive and whitespace-stripped."""
+    monkeypatch.setenv(_ENV_VAR, env_value)
</code_context>

<issue_to_address>
**suggestion (testing):** Add a unit test for `CUPRUM_STREAM_BACKEND=auto` when Rust is available

The parametrized `test_env_var_case_insensitive` covers `AUTO`/`auto` only in the "Rust unavailable" case (since `expected` is never `StreamBackend.RUST`). The only coverage for auto+Rust-available is currently via the *unset* env var path (`test_auto_returns_rust_when_available`). Please add a test (or parametrized row) for `CUPRUM_STREAM_BACKEND="auto"` with `is_available` returning `True`, and assert that `get_stream_backend()` returns `StreamBackend.RUST`.

```suggestion
@pytest.mark.parametrize(
    ("env_value", "expected"),
    [
        ("AUTO", StreamBackend.PYTHON),
        ("Rust", StreamBackend.RUST),
        ("PYTHON", StreamBackend.PYTHON),
        ("  auto  ", StreamBackend.PYTHON),
        ("auto", StreamBackend.RUST),
    ],
    ids=[
        "upper-auto",
        "mixed-rust",
        "upper-python",
        "whitespace-auto",
        "lower-auto-rust-available",
    ],
)
def test_env_var_case_insensitive(
    monkeypatch: pytest.MonkeyPatch,
    env_value: str,
    expected: StreamBackend,
) -> None:
```
</issue_to_address>

### Comment 4
<location> `cuprum/unittests/test_backend.py:99-110` </location>
<code_context>
+# -- env var validation -------------------------------------------------------
+
+
+def test_invalid_env_var_raises_value_error(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """An unrecognised env var value raises ValueError."""
+    monkeypatch.setenv(_ENV_VAR, "turbo")
+    monkeypatch.setattr(_rust_backend, "is_available", lambda: False)
+
+    with pytest.raises(ValueError, match="turbo"):
+        get_stream_backend()
+
</code_context>

<issue_to_address>
**suggestion (testing):** Consider adding a test for `CUPRUM_STREAM_BACKEND` set to an empty string

`_read_backend_env` currently treats both unset and empty `CUPRUM_STREAM_BACKEND` as `AUTO` via `.get(..., '').strip().lower()` and `if not raw`. We already cover the unset case and invalid non-empty values (e.g. `turbo`), but not the explicit empty-string case. Please add a test that does `monkeypatch.setenv(_ENV_VAR, "")` and asserts that `get_stream_backend()` behaves like auto mode (respecting Rust availability).

```suggestion
# -- env var validation -------------------------------------------------------


def test_invalid_env_var_raises_value_error(
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """An unrecognised env var value raises ValueError."""
    monkeypatch.setenv(_ENV_VAR, "turbo")
    monkeypatch.setattr(_rust_backend, "is_available", lambda: False)

    with pytest.raises(ValueError, match="turbo"):
        get_stream_backend()


def test_empty_env_var_uses_auto_mode(
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """An explicit empty env var value behaves like AUTO."""
    monkeypatch.setenv(_ENV_VAR, "")

    for available, expected in (
        (True, StreamBackend.RUST),
        (False, StreamBackend.PYTHON),
    ):
        monkeypatch.setattr(
            _rust_backend,
            "is_available",
            lambda available=available: available,
        )
        assert get_stream_backend() is expected
```
</issue_to_address>

### Comment 5
<location> `cuprum/unittests/test_backend.py:90-96` </location>
<code_context>
+# -- forced python mode -------------------------------------------------------
+
+
+def test_forced_python_returns_python(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Forced Python mode returns PYTHON regardless of availability."""
+    monkeypatch.setenv(_ENV_VAR, "python")
+    # Availability should not matter — do not monkeypatch is_available.
+    assert get_stream_backend() is StreamBackend.PYTHON
+
</code_context>

<issue_to_address>
**suggestion (testing):** Strengthen the forced-Python test by asserting the availability probe is not used

Since this test is meant to prove that availability is ignored in forced-Python mode, it should also verify that `_rust_backend.is_available` is never called. Consider monkeypatching `_rust_backend.is_available` (e.g., to raise or increment a counter) and then asserting it was not invoked. This ensures the behavior isn’t just coincidentally returning `PYTHON` while still consulting the availability probe.

```suggestion
def test_forced_python_returns_python(
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Forced Python mode returns PYTHON without consulting availability."""
    # If `is_available` is ever called, this test should fail.
    called = False

    def _is_available_sentinel() -> bool:
        nonlocal called
        called = True
        raise AssertionError(
            "_rust_backend.is_available must not be called in forced Python mode"
        )

    monkeypatch.setattr(_rust_backend, "is_available", _is_available_sentinel)
    monkeypatch.setenv(_ENV_VAR, "python")

    backend = get_stream_backend()
    assert backend is StreamBackend.PYTHON
    assert called is False
```
</issue_to_address>

### Comment 6
<location> `docs/execplans/4-2-4-backend-dispatcher.md:341` </location>
<code_context>
+Each test clears the `_check_rust_available` cache in an autouse fixture to
+prevent cross-test pollution.
+
+### Stage C: Add BDD feature file and behavioural tests
+
+Create `tests/features/backend_dispatcher.feature`:
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The acronym “BDD” is used here without being defined on first use, which violates the acronym-definition requirement.

Please expand the acronym on first use, for example: “Add behaviour-driven development (BDD) feature file and behavioural tests.” Subsequent uses can then use “BDD” on its own.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 7
<location> `docs/execplans/4-2-4-backend-dispatcher.md:475` </location>
<code_context>
+- Documentation: `make markdownlint` passes. Roadmap item 4.2.4 is marked
+  done.
+
+**Quality method (how we check):**
+
+1. Run `make check-fmt && make typecheck && make lint && make test`
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The phrase “how we check” uses the first-person pronoun “we”, which the style guidelines prohibit.

Consider rephrasing to avoid first-person pronouns, for example: “Quality method (how checks are performed):” or “Quality method (how verification is performed):”.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Avoid 2nd person or 1st person pronouns ("I", "you", "we").

</details>
</issue_to_address>

### Comment 8
<location> `docs/execplans/4-2-4-backend-dispatcher.md:351` </location>
<code_context>
+      availability. The dispatcher caches the availability check for
+      performance.
+
+      Scenario: Auto mode selects Python when Rust is unavailable
+        Given the Rust extension is not available
+        And the stream backend environment variable is unset
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The Gherkin steps in this example use the first-person pronoun “I”, which is not allowed by the style rules.

Within the Gherkin example under this scenario (and the subsequent scenarios), steps such as “When I resolve the stream backend” and “When I attempt to resolve the stream backend” use the pronoun “I”. Please rephrase them to avoid first-person, for example: “When the stream backend is resolved” or “When an attempt is made to resolve the stream backend”. Apply the same adjustment to all similar steps in the listed scenarios.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Avoid 2nd person or 1st person pronouns ("I", "you", "we").

</details>
</issue_to_address>
~~~

File: cuprum/_backend.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        if _check_rust_available():
+            return StreamBackend.RUST
+        msg = (
+            f"Rust stream backend requested via {_ENV_VAR}=rust "
+            "but the Rust extension is not available"
+        )
+        raise ImportError(msg)
+
+    # AUTO: prefer Rust when available, otherwise Python.
+    if _check_rust_available():
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/_backend.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/execplans/4-2-4-backend-dispatcher.md [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+  - Likelihood: certain (by design — caching is a stated requirement)
+  - Mitigation: Used `functools.lru_cache(maxsize=1)` on
+    `_check_rust_available()` inside `_backend.py`. ✅ Resolved.
+
+- **Risk**: Environment variable could be set to an unexpected value (mixed
+  case, whitespace, empty string).
+  - Severity: low
+  - Likelihood: medium
+  - Mitigation: `_read_backend_env()` strips and lowercases the value; raises
+    `ValueError` for unrecognised values. ✅ Resolved.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/execplans/4-2-4-backend-dispatcher.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/execplans/4-2-4-backend-dispatcher.md [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+Create a new module at `cuprum/_backend.py` containing:
+
+1. **`StreamBackend` enum** — a `StrEnum` with three members: `AUTO`,
+   `RUST`, `PYTHON`. The `auto()` values will produce lowercase strings
+   `"auto"`, `"rust"`, `"python"`.
+
+2. **`_read_backend_env() -> StreamBackend`** — reads `CUPRUM_STREAM_BACKEND`
+   from `os.environ`, strips whitespace, lowercases, defaults to `"auto"`.
+   Raises `ValueError` for unrecognised values with a message listing valid
+   options.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/execplans/4-2-4-backend-dispatcher.md:279]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
